src/librustc_mir/transform/simplify_comparison_integral.rs

```diff
@@ -0,0 +1,252 @@
+            IntTy::I128 => scalar
+                .to_i128()
+                .map_err(err_tracer)
+                .ok()
+                .map(|x| truncate(x as u128, Size::from_bits(128))),
+        },
+        Uint(int) => match int {
+            UintTy::Usize => {
+                scalar.to_machine_usize(&tcx).map_err(err_tracer).ok().map(|x| x as u128)
+            }
+            UintTy::U8 => scalar.to_u8().map_err(err_tracer).ok().map(|x| x as u128),
+            UintTy::U16 => scalar.to_u16().map_err(err_tracer).ok().map(|x| x as u128),
+            UintTy::U32 => scalar.to_u32().map_err(err_tracer).ok().map(|x| x as u128),
+            UintTy::U64 => scalar.to_u64().map_err(err_tracer).ok().map(|x| x as u128),
+            UintTy::U128 => scalar.to_u128().map_err(err_tracer).ok().map(|x| x as u128),
+        },
+        _ => None,
+    }
+}
+
+struct OptimizationFinder<'a, 'tcx> {
+    body: &'a Body<'tcx>,
+}
+
+impl<'a, 'tcx> OptimizationFinder<'a, 'tcx> {
+    fn find_optimizations(&self) -> Vec<OptimizationInfo<'tcx>> {
+        self.body
+            .basic_blocks()
+            .iter_enumerated()
+            .filter_map(|(bb_idx, bb)| {
+                // find switch
+                let (place_switched_on, values, targets) = match &bb.terminator().kind {
+                    rustc_middle::mir::TerminatorKind::SwitchInt {
+                        discr, values, targets, ..
+                    } => Some((discr.place()?, values, targets)),
+                    _ => None,
+                }?;
+
+                // find the statement that assigns the place being switched on
+                bb.statements.iter().enumerate().rev().find_map(|(stmt_idx, stmt)| {
+                    match &stmt.kind {
+                        rustc_middle::mir::StatementKind::Assign(box (lhs, rhs))
+                            if *lhs == place_switched_on =>
+                        {
+                            match rhs {
+                                Rvalue::BinaryOp(op, left, right)
+                                    if *op == BinOp::Eq || *op == BinOp::Ne =>
+                                {
+                                    // check that either left or right is a constant.
+                                    // if any are, we can use the other to switch on, and the constant as a value in a switch
+                                    use Operand::*;
+                                    match (left, right) {
+                                        (
+                                            Constant(branch_value),
+                                            Copy(to_switch_on) | Move(to_switch_on),
+                                        )
+                                        | (
+                                            Copy(to_switch_on) | Move(to_switch_on),
+                                            Constant(branch_value),
+                                        ) => {
```

Comment:
Move this into a function so you can call it and use `?` to skip the case where this pattern isn't matched.

Reply:
with "move this" I meant the pattern and the extraction of the fiels from the pattern